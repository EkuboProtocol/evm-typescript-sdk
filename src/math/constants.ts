export const MAX_U256 =
  0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffn;
export const MAX_U128 = 0xffffffffffffffffffffffffffffffffn;
