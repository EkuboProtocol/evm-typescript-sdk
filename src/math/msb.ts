const MAX = 1n << 128n;

export default function msb(x: bigint): number {
  if (x >= MAX) {
    throw new Error("x too large");
  }
  if (x <= 0n) {
    throw new Error("x must be positive");
  }
  let res = 0;
  if (x >= 0x10000000000000000n) {
    x = x >> 64n;
    res += 64;
  }
  if (x >= 0x100000000n) {
    x = x >> 32n;
    res += 32;
  }
  if (x >= 0x10000n) {
    x = x >> 16n;
    res += 16;
  }
  if (x >= 0x100n) {
    x = x >> 8n;
    res += 8;
  }
  if (x >= 0x10n) {
    x = x >> 4n;
    res += 4;
  }
  if (x >= 0x04n) {
    x = x >> 2n;
    res += 2;
  }
  if (x >= 0x02n) {
    res += 1;
  }
  return res;
}
