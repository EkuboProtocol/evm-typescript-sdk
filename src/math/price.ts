import { MAX_U256 } from "./constants";

export function nextSqrtRatioFromAmount0(
  sqrtRatio: bigint,
  liquidity: bigint,
  amount0: bigint,
): bigint | null {
  if (amount0 === 0n) return sqrtRatio;

  if (liquidity === 0n) throw new Error("NO_LIQUIDITY");

  const numerator1 = liquidity << 128n;

  // because quotient is rounded down, this price movement is also rounded towards sqrt_ratio
  if (amount0 < 0n) {
    const product = amount0 * -1n * sqrtRatio;

    if (product >= MAX_U256) {
      return null;
    }

    const denominator = numerator1 - product;

    if (denominator < 0n) {
      return null;
    }

    const num = numerator1 * sqrtRatio;

    const result = num / denominator + (num % denominator === 0n ? 0n : 1n);

    if (result > MAX_U256) {
      return null;
    }

    return result;
  } else {
    const denomP1 = numerator1 / sqrtRatio;

    const denom = denomP1 + amount0;
    const quotient = numerator1 / denom;
    const remainder = numerator1 % denom;

    if (remainder === 0n) return quotient;
    const sum = quotient + 1n;
    if (sum > MAX_U256) return null;
    return sum;
  }
}

export function nextSqrtRatioFromAmount1(
  sqrtRatio: bigint,
  liquidity: bigint,
  amount1: bigint,
): bigint | null {
  if (amount1 === 0n) return sqrtRatio;

  if (liquidity === 0n) throw new Error("NO_LIQUIDITY");

  const amountShifted = amount1 << 128n;
  const quotient = amountShifted / liquidity;
  const remainder = amountShifted % liquidity;

  // because quotient is rounded down, this price movement is also rounded towards sqrt_ratio
  if (amount1 < 0n) {
    // adding amount1, taking out amount0
    const res = sqrtRatio + quotient;
    if (res < 0n) {
      return null;
    }
    if (remainder === 0n) {
      return res;
    } else {
      if (res != 0n) {
        return res - 1n;
      } else {
        return null;
      }
    }
  } else {
    // adding amount1, taking out amount0, price goes up
    const res = sqrtRatio + quotient;
    if (res > MAX_U256) {
      return null;
    } else {
      return res;
    }
  }
}
