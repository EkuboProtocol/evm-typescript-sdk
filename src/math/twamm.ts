import { MAX_U256 } from "./constants";

// ~ ln(2**128) * 2**64
export const EXPONENT_LIMIT: bigint = 1623313478486440542208n;

export function calculateNextSqrtRatio(
  sqrtRatio: bigint,
  liquidity: bigint,
  token0SaleRate: bigint,
  token1SaleRate: bigint,
  timeElapsed: bigint,
  fee: bigint,
): bigint {
  let sqrtSaleRatio = sqrt((token1SaleRate << 128n) / token0SaleRate) << 64n;

  if (liquidity === 0n) {
    return sqrtSaleRatio;
  }

  const sRate =
    (sqrt(token1SaleRate * token0SaleRate) * ((1n << 128n) - fee)) /
    (1n << 128n);

  const roundUp = sqrtRatio > sqrtSaleRatio;

  const exponent = div(0x200000000n * timeElapsed * sRate, liquidity, roundUp);

  if (exponent > EXPONENT_LIMIT) {
    return sqrtSaleRatio;
  }

  const e = exp(exponent);

  const [num, sign] = roundUp
    ? [sqrtRatio - sqrtSaleRatio, true]
    : [sqrtSaleRatio - sqrtRatio, false];

  const c = div(num << 128n, sqrtSaleRatio + sqrtRatio, roundUp);

  const [term1, term2] = [e - c, e + c];
  const scale = sign
    ? div(term2 << 128n, term1, roundUp)
    : div(term1 << 128n, term2, roundUp);

  return (sqrtSaleRatio * scale) >> 128n;
}

function div(x: bigint, y: bigint, round: boolean): bigint {
  const quotient = x / y;
  const remainder: bigint = x % y;
  return quotient + (remainder !== 0n && round ? 1n : 0n);
}

// Computes e^x where x is a fixed point 64.64 number and the result is a fixed point 128.128 number
export function exp(x: bigint): bigint {
  if (x >= 0x20000000000000000) {
    let half = exp(x / 2n);
    return (half * half) >> 128n;
  } else {
    return expInner(x);
  }
}

// Computes e^x where x is a fixed point 64.64 number that is less than the real number 2
function expInner(x: bigint): bigint {
  if (x >= 0x20000000000000000n) {
    throw new Error("Invalid input");
  }

  let ratio = 0x100000000000000000000000000000000n;

  if ((x & 0x1n) != 0n) {
    ratio = 0xffffffffffffffff0000000000000000n;
  }
  if ((x & 0x2n) != 0n) {
    ratio = (ratio * 0xfffffffffffffffe0000000000000002n) >> 128n;
  }
  if ((x & 0x4n) != 0n) {
    ratio = (ratio * 0xfffffffffffffffc0000000000000008n) >> 128n;
  }
  if ((x & 0x8n) != 0n) {
    ratio = (ratio * 0xfffffffffffffff80000000000000020n) >> 128n;
  }
  if ((x & 0x10n) != 0n) {
    ratio = (ratio * 0xfffffffffffffff00000000000000080n) >> 128n;
  }
  if ((x & 0x20n) != 0n) {
    ratio = (ratio * 0xffffffffffffffe00000000000000200n) >> 128n;
  }
  if ((x & 0x40n) != 0n) {
    ratio = (ratio * 0xffffffffffffffc00000000000000800n) >> 128n;
  }
  if ((x & 0x80n) != 0n) {
    ratio = (ratio * 0xffffffffffffff800000000000002000n) >> 128n;
  }
  if ((x & 0x100n) != 0n) {
    ratio = (ratio * 0xffffffffffffff000000000000008000n) >> 128n;
  }
  if ((x & 0x200n) != 0n) {
    ratio = (ratio * 0xfffffffffffffe000000000000020000n) >> 128n;
  }
  if ((x & 0x400n) != 0n) {
    ratio = (ratio * 0xfffffffffffffc000000000000080000n) >> 128n;
  }
  if ((x & 0x800n) != 0n) {
    ratio = (ratio * 0xfffffffffffff8000000000000200000n) >> 128n;
  }
  if ((x & 0x1000n) != 0n) {
    ratio = (ratio * 0xfffffffffffff0000000000000800000n) >> 128n;
  }
  if ((x & 0x2000n) != 0n) {
    ratio = (ratio * 0xffffffffffffe0000000000002000000n) >> 128n;
  }
  if ((x & 0x4000n) != 0n) {
    ratio = (ratio * 0xffffffffffffc0000000000008000000n) >> 128n;
  }
  if ((x & 0x8000n) != 0n) {
    ratio = (ratio * 0xffffffffffff80000000000020000000n) >> 128n;
  }
  if ((x & 0x10000n) != 0n) {
    ratio = (ratio * 0xffffffffffff00000000000080000000n) >> 128n;
  }
  if ((x & 0x20000n) != 0n) {
    ratio = (ratio * 0xfffffffffffe00000000000200000000n) >> 128n;
  }
  if ((x & 0x40000n) != 0n) {
    ratio = (ratio * 0xfffffffffffc00000000000800000000n) >> 128n;
  }
  if ((x & 0x80000n) != 0n) {
    ratio = (ratio * 0xfffffffffff800000000002000000000n) >> 128n;
  }
  if ((x & 0x100000n) != 0n) {
    ratio = (ratio * 0xfffffffffff000000000008000000000n) >> 128n;
  }
  if ((x & 0x200000n) != 0n) {
    ratio = (ratio * 0xffffffffffe000000000020000000000n) >> 128n;
  }
  if ((x & 0x400000n) != 0n) {
    ratio = (ratio * 0xffffffffffc00000000007ffffffffffn) >> 128n;
  }
  if ((x & 0x800000n) != 0n) {
    ratio = (ratio * 0xffffffffff80000000001ffffffffffbn) >> 128n;
  }
  if ((x & 0x1000000n) != 0n) {
    ratio = (ratio * 0xffffffffff00000000007fffffffffd5n) >> 128n;
  }
  if ((x & 0x2000000n) != 0n) {
    ratio = (ratio * 0xfffffffffe0000000001fffffffffeabn) >> 128n;
  }
  if ((x & 0x4000000n) != 0n) {
    ratio = (ratio * 0xfffffffffc0000000007fffffffff555n) >> 128n;
  }
  if ((x & 0x8000000n) != 0n) {
    ratio = (ratio * 0xfffffffff8000000001fffffffffaaabn) >> 128n;
  }
  if ((x & 0x10000000n) != 0n) {
    ratio = (ratio * 0xfffffffff0000000007ffffffffd5555n) >> 128n;
  }
  if ((x & 0x20000000n) != 0n) {
    ratio = (ratio * 0xffffffffe000000001ffffffffeaaaabn) >> 128n;
  }
  if ((x & 0x40000000n) != 0n) {
    ratio = (ratio * 0xffffffffc000000007ffffffff555555n) >> 128n;
  }
  if ((x & 0x80000000n) != 0n) {
    ratio = (ratio * 0xffffffff800000001ffffffffaaaaaabn) >> 128n;
  }
  if ((x & 0x100000000n) != 0n) {
    ratio = (ratio * 0xffffffff000000007fffffffd5555555n) >> 128n;
  }
  if ((x & 0x200000000n) != 0n) {
    ratio = (ratio * 0xfffffffe00000001fffffffeaaaaaaabn) >> 128n;
  }
  if ((x & 0x400000000n) != 0n) {
    ratio = (ratio * 0xfffffffc00000007fffffff555555560n) >> 128n;
  }
  if ((x & 0x800000000n) != 0n) {
    ratio = (ratio * 0xfffffff80000001fffffffaaaaaaab55n) >> 128n;
  }
  if ((x & 0x1000000000n) != 0n) {
    ratio = (ratio * 0xfffffff00000007ffffffd5555556000n) >> 128n;
  }
  if ((x & 0x2000000000n) != 0n) {
    ratio = (ratio * 0xffffffe0000001ffffffeaaaaaab5555n) >> 128n;
  }
  if ((x & 0x4000000000n) != 0n) {
    ratio = (ratio * 0xffffffc0000007ffffff555555600000n) >> 128n;
  }
  if ((x & 0x8000000000n) != 0n) {
    ratio = (ratio * 0xffffff8000001ffffffaaaaaab555555n) >> 128n;
  }
  if ((x & 0x10000000000n) != 0n) {
    ratio = (ratio * 0xffffff0000007fffffd555555ffffffen) >> 128n;
  }
  if ((x & 0x20000000000n) != 0n) {
    ratio = (ratio * 0xfffffe000001fffffeaaaaab55555511n) >> 128n;
  }
  if ((x & 0x40000000000n) != 0n) {
    ratio = (ratio * 0xfffffc000007fffff555555ffffff777n) >> 128n;
  }
  if ((x & 0x80000000000n) != 0n) {
    ratio = (ratio * 0xfffff800001fffffaaaaab5555544444n) >> 128n;
  }
  if ((x & 0x100000000000n) != 0n) {
    ratio = (ratio * 0xfffff000007ffffd55555fffffddddden) >> 128n;
  }
  if ((x & 0x200000000000n) != 0n) {
    ratio = (ratio * 0xffffe00001ffffeaaaab555551111128n) >> 128n;
  }
  if ((x & 0x400000000000n) != 0n) {
    ratio = (ratio * 0xffffc00007ffff55555fffff77777d28n) >> 128n;
  }
  if ((x & 0x800000000000n) != 0n) {
    ratio = (ratio * 0xffff80001ffffaaaab5555444445b05bn) >> 128n;
  }
  if ((x & 0x1000000000000n) != 0n) {
    ratio = (ratio * 0xffff00007fffd5555ffffdddde38e381n) >> 128n;
  }
  if ((x & 0x2000000000000n) != 0n) {
    ratio = (ratio * 0xfffe0001fffeaaab5555111127d276a7n) >> 128n;
  }
  if ((x & 0x4000000000000n) != 0n) {
    ratio = (ratio * 0xfffc0007fff5555ffff7777d27cf3cf5n) >> 128n;
  }
  if ((x & 0x8000000000000n) != 0n) {
    ratio = (ratio * 0xfff8001fffaaab55544445b0596597f9n) >> 128n;
  }
  if ((x & 0x10000000000000n) != 0n) {
    ratio = (ratio * 0xfff0007ffd555fffddde38e2be2d82d5n) >> 128n;
  }
  if ((x & 0x20000000000000n) != 0n) {
    ratio = (ratio * 0xffe001ffeaab55511127d21522f2295cn) >> 128n;
  }
  if ((x & 0x40000000000000n) != 0n) {
    ratio = (ratio * 0xffc007ff555fff777d279e7b87acece0n) >> 128n;
  }
  if ((x & 0x80000000000000n) != 0n) {
    ratio = (ratio * 0xff801ffaab554445b04105b043e8f48dn) >> 128n;
  }
  if ((x & 0x100000000000000n) != 0n) {
    ratio = (ratio * 0xff007fd55ffdde38d68f08c257e0ce3fn) >> 128n;
  }
  if ((x & 0x200000000000000n) != 0n) {
    ratio = (ratio * 0xfe01feab551127cbfe5f89994c44216fn) >> 128n;
  }
  if ((x & 0x400000000000000n) != 0n) {
    ratio = (ratio * 0xfc07f55ff77d2493e885eeaa756ad523n) >> 128n;
  }
  if ((x & 0x800000000000000n) != 0n) {
    ratio = (ratio * 0xf81fab5445aebc8a58055fcbbb139ae9n) >> 128n;
  }
  if ((x & 0x1000000000000000n) != 0n) {
    ratio = (ratio * 0xf07d5fde38151e72f18ff03049ac5d7fn) >> 128n;
  }
  if ((x & 0x2000000000000000n) != 0n) {
    ratio = (ratio * 0xe1eb51276c110c3c3eb1269f2f5d4afbn) >> 128n;
  }
  if ((x & 0x4000000000000000n) != 0n) {
    ratio = (ratio * 0xc75f7cf564105743415cbc9d6368f3b9n) >> 128n;
  }
  if ((x & 0x8000000000000000n) != 0n) {
    ratio = (ratio * 0x9b4597e37cb04ff3d675a35530cdd768n) >> 128n;
  }
  if ((x & 0x10000000000000000n) != 0n) {
    ratio = (ratio * 0x5e2d58d8b3bcdf1abadec7829054f90en) >> 128n;
  }

  if (x != 0n) {
    ratio = MAX_U256 / ratio;
  }

  return ratio;
}

export function sqrt(x: bigint): bigint {
  if (x < 0n) {
    throw new Error("Square root of negative numbers is not supported.");
  }
  if (x < 2n) {
    return x;
  }

  let x0: bigint = x / 2n;
  let x1: bigint = (x0 + x / x0) / 2n;

  while (x0 > x1) {
    x0 = x1;
    x1 = (x0 + x / x0) / 2n;
  }

  return x0;
}
