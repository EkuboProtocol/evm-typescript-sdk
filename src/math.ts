import { amount0Delta, amount1Delta } from "./math/delta";
import msb from "./math/msb";
import {
  nextSqrtRatioFromAmount0,
  nextSqrtRatioFromAmount1,
} from "./math/price";
import { computeStep } from "./math/swap";
import { toSqrtRatio } from "./math/tick";
import { calculateNextSqrtRatio } from "./math/twamm";

export {
  amount0Delta,
  amount1Delta,
  msb,
  nextSqrtRatioFromAmount0,
  nextSqrtRatioFromAmount1,
  computeStep,
  toSqrtRatio,
  calculateNextSqrtRatio,
};
