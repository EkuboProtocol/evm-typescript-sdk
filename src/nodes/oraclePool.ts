import { MAX_TICK_SPACING } from "../math/tick";
import { BasePool } from "./basePool";
import {
  BasePoolResources,
  BasePoolState,
  NodeKey,
  Quote,
  QuoteNode,
  QuoteParams,
  Tick,
} from "./quoteNode";

export interface OracleResources extends BasePoolResources {
  readonly snapshotUpdated: boolean;
}

export interface OraclePoolState extends BasePoolState {
  readonly lastSnapshotTime: bigint;
}

export class OraclePool implements QuoteNode<OracleResources, OraclePoolState> {
  private readonly extension: bigint;
  private readonly basePool: BasePool;
  private readonly lastSnapshotTime: bigint;

  constructor({
    token0,
    token1,
    sqrtRatio,
    liquidity,
    tick,
    sortedTicks,
    extension,
    lastSnapshotTime,
  }: {
    token0: bigint;
    token1: bigint;
    extension: bigint;
    sqrtRatio: bigint;
    liquidity: bigint;
    tick: number;
    sortedTicks: Tick[];
    lastSnapshotTime: bigint;
  }) {
    this.basePool = new BasePool({
      token0,
      token1,
      fee: 0n,
      tickSpacing: MAX_TICK_SPACING,
      sortedTicks,
      tick,
      sqrtRatio,
      liquidity,
    });
    this.extension = extension;
    this.lastSnapshotTime = lastSnapshotTime;
  }

  public get key(): NodeKey {
    return { ...this.basePool.key, extension: this.extension };
  }

  public get state(): Readonly<OraclePoolState> {
    return { ...this.basePool.state, lastSnapshotTime: this.lastSnapshotTime };
  }

  public get sortedTicks(): Tick[] {
    return this.basePool.sortedTicks;
  }

  quote(
    params: QuoteParams<OraclePoolState>
  ): Quote<OracleResources, OraclePoolState> {
    const quote = this.basePool.quote(params);

    const state = params.overrideState ?? this.state;
    const blockTimestamp = BigInt(params.meta.block.time);
    const snapshotUpdated = state.lastSnapshotTime < blockTimestamp;

    return {
      ...quote,
      executionResources: {
        ...quote.executionResources,
        snapshotUpdated,
      },
      stateAfter: {
        ...state,
        lastSnapshotTime: snapshotUpdated
          ? blockTimestamp
          : state.lastSnapshotTime,
      },
    };
  }

  combineResources(
    resource: OracleResources,
    additionalResources: OracleResources
  ): OracleResources {
    return {
      ...this.basePool.combineResources(resource, additionalResources),
      snapshotUpdated:
        resource.snapshotUpdated || additionalResources.snapshotUpdated,
    };
  }

  initialResources(): OracleResources {
    return {
      ...this.basePool.initialResources(),
      snapshotUpdated: false,
    };
  }

  hasLiquidity(): boolean {
    return this.basePool.hasLiquidity();
  }
}
