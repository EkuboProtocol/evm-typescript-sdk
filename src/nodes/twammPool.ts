import { MAX_SQRT_RATIO, MAX_TICK_SPACING, MIN_SQRT_RATIO } from "../math/tick";
import { calculateNextSqrtRatio } from "../math/twamm";
import { BasePool } from "./basePool";
import {
  BasePoolResources,
  BasePoolState,
  NodeKey,
  Quote,
  QuoteNode,
  QuoteParams,
  Tick,
} from "./quoteNode";

export const MAX_BOUND_USABLE_TICK_MAGNITUDE = 88368108;
const MAX_BOUNDS_MIN_SQRT_RATIO: bigint = 22027144413679976675n;
const MAX_BOUNDS_MAX_SQRT_RATIO: bigint =
  5256790760649093508123362461711849782692726119655358142129n;

export interface TwammResources extends BasePoolResources {
  virtualOrderSecondsExecuted: number;
  virtualOrderDeltaTimesCrossed: number;
}

export interface TwammSaleRateDelta {
  saleRateDelta0: bigint;
  saleRateDelta1: bigint;
  time: number;
}

export interface TwammPoolState extends BasePoolState {
  readonly token0SaleRate: bigint;
  readonly token1SaleRate: bigint;
  readonly lastExecutionTime: number;
}

export class TwammPool implements QuoteNode<TwammResources, TwammPoolState> {
  public get key(): NodeKey {
    return { ...this.basePool.key, extension: this.extension };
  }

  public get sortedTicks(): Tick[] {
    return this.basePool.sortedTicks;
  }

  private readonly extension: bigint;
  readonly basePool: BasePool;

  // state
  public readonly token0SaleRate: bigint;
  public readonly token1SaleRate: bigint;
  public readonly lastExecutionTime: number;
  public readonly saleRateDeltas: Readonly<TwammSaleRateDelta[]>;

  constructor({
    token0,
    token1,
    fee,
    extension,

    sqrtRatio,
    liquidity,
    tick,

    token0SaleRate,
    token1SaleRate,
    lastExecutionTime,
    saleRateDeltas,
  }: {
    token0: bigint;
    token1: bigint;
    fee: bigint;
    extension: bigint;

    sqrtRatio: bigint;
    liquidity: bigint;
    tick: number;

    token0SaleRate: bigint;
    token1SaleRate: bigint;
    lastExecutionTime: number;
    saleRateDeltas: TwammSaleRateDelta[];
  }) {
    this.extension = extension;
    this.basePool = new BasePool({
      token0,
      token1,
      fee,
      sqrtRatio,
      liquidity,
      tick,
      tickSpacing: MAX_TICK_SPACING,
      sortedTicks: [
        { tick: -MAX_BOUND_USABLE_TICK_MAGNITUDE, liquidityDelta: liquidity },
        { tick: MAX_BOUND_USABLE_TICK_MAGNITUDE, liquidityDelta: -liquidity },
      ],
    });

    this.token0SaleRate = token0SaleRate;
    this.token1SaleRate = token1SaleRate;
    this.lastExecutionTime = lastExecutionTime;
    this.saleRateDeltas = saleRateDeltas;
  }

  initialResources(): TwammResources {
    return {
      ...this.basePool.initialResources(),
      virtualOrderSecondsExecuted: 0,
      virtualOrderDeltaTimesCrossed: 0,
    };
  }

  combineResources(
    resource: TwammResources,
    additionalResources: TwammResources,
  ): TwammResources {
    return {
      ...this.basePool.combineResources(resource, additionalResources),
      virtualOrderDeltaTimesCrossed:
        resource.virtualOrderDeltaTimesCrossed +
        additionalResources.virtualOrderDeltaTimesCrossed,
      virtualOrderSecondsExecuted:
        resource.virtualOrderSecondsExecuted +
        additionalResources.virtualOrderSecondsExecuted,
    };
  }

  public quote({
    tokenAmount,
    sqrtRatioLimit,
    overrideState,
    meta,
  }: QuoteParams<TwammPoolState>): Quote<TwammResources, TwammPoolState> {
    const {
      block: { time: currentTime },
    } = meta;

    const initialState = overrideState ?? this.state;

    let {
      sqrtRatio: nextSqrtRatio,
      token0SaleRate,
      token1SaleRate,
      lastExecutionTime,
    } = initialState;

    const virtualOrderSecondsExecuted = currentTime - lastExecutionTime;
    if (virtualOrderSecondsExecuted < 0)
      throw new Error("Last execution time exceeds block time");

    let virtualOrderDeltaTimesCrossed: number = 0;

    let nextSaleRateDeltaIndex = this.saleRateDeltas.findIndex(
      (srd) => srd.time > lastExecutionTime,
    );

    // this is the current state of the base pool during the iteration
    let basePoolStateOverride: BasePoolState | undefined = overrideState;
    let basePoolExecutionResources: BasePoolResources =
      this.basePool.initialResources();

    while (lastExecutionTime !== currentTime) {
      const saleRateDelta = this.saleRateDeltas[nextSaleRateDeltaIndex];
      const nextExecutionTime = saleRateDelta
        ? Math.min(saleRateDelta.time, currentTime)
        : currentTime;

      const timeElapsed = BigInt(nextExecutionTime - lastExecutionTime);

      const [amount0, amount1] = [
        (token0SaleRate * timeElapsed) >> 32n,
        (token1SaleRate * timeElapsed) >> 32n,
      ];

      if (amount0 > 0n && amount1 > 0n) {
        const currentSqrtRatio = max(
          MAX_BOUNDS_MIN_SQRT_RATIO,
          min(MAX_BOUNDS_MAX_SQRT_RATIO, nextSqrtRatio),
        );

        nextSqrtRatio = calculateNextSqrtRatio(
          currentSqrtRatio,
          // this is what the twamm pool uses for liquidity for its swap calc and it cannot be overridden
          this.basePool.sortedTicks[0].liquidityDelta,
          token0SaleRate,
          token1SaleRate,
          timeElapsed,
          this.key.fee,
        );

        const [token, amount] =
          currentSqrtRatio < nextSqrtRatio
            ? [this.basePool.key.token1, amount1]
            : [this.basePool.key.token0, amount0];

        const quote = this.basePool.quote({
          tokenAmount: {
            amount,
            token,
          },
          sqrtRatioLimit: nextSqrtRatio,
          meta,
          overrideState: basePoolStateOverride,
        });

        basePoolStateOverride = quote.stateAfter;
        basePoolExecutionResources = this.basePool.combineResources(
          basePoolExecutionResources,
          quote.executionResources,
        );
      } else if (amount0 > 0n || amount1 > 0n) {
        const [amount, isToken1, sqrtRatioLimit] =
          amount0 !== 0n
            ? [amount0, false, MIN_SQRT_RATIO]
            : [amount1, true, MAX_SQRT_RATIO];

        const quote = this.basePool.quote({
          tokenAmount: {
            amount,
            token: isToken1
              ? this.basePool.key.token1
              : this.basePool.key.token0,
          },
          sqrtRatioLimit,
          meta,
          overrideState: basePoolStateOverride,
        });

        basePoolStateOverride = quote.stateAfter;
        basePoolExecutionResources = this.basePool.combineResources(
          basePoolExecutionResources,
          quote.executionResources,
        );

        nextSqrtRatio = basePoolStateOverride.sqrtRatio;
      }

      // if we executed up to the next sale rate delta, we need to apply the delta
      if (nextExecutionTime === saleRateDelta?.time) {
        token0SaleRate += saleRateDelta.saleRateDelta0;
        token1SaleRate += saleRateDelta.saleRateDelta1;
        nextSaleRateDeltaIndex++;
        virtualOrderDeltaTimesCrossed++;
      }

      lastExecutionTime = nextExecutionTime;
    }

    const {
      consumedAmount,
      calculatedAmount,
      executionResources,
      stateAfter,
      isPriceIncreasing,
      feesPaid,
    } = this.basePool.quote({
      tokenAmount,
      sqrtRatioLimit,
      meta,
      overrideState: basePoolStateOverride,
    });

    return {
      isPriceIncreasing,
      consumedAmount,
      calculatedAmount,
      feesPaid,
      executionResources: {
        ...this.basePool.combineResources(
          basePoolExecutionResources,
          executionResources,
        ),
        virtualOrderSecondsExecuted,
        virtualOrderDeltaTimesCrossed,
      },
      stateAfter: {
        ...stateAfter,
        token0SaleRate,
        token1SaleRate,
        lastExecutionTime: currentTime,
      },
    };
  }

  public hasLiquidity(): boolean {
    return this.basePool.hasLiquidity();
  }

  get state(): Readonly<TwammPoolState> {
    return {
      ...this.basePool.state,
      token0SaleRate: this.token0SaleRate,
      token1SaleRate: this.token1SaleRate,
      lastExecutionTime: this.lastExecutionTime,
    };
  }
}

function min(x: bigint, y: bigint): bigint {
  return x > y ? y : x;
}

function max(x: bigint, y: bigint): bigint {
  return x > y ? x : y;
}
