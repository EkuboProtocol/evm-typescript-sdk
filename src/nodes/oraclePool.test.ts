import { toSqrtRatio } from "../math/tick";
import { OraclePool } from "./oraclePool";

import { describe, expect, it } from "vitest";
import { MAX_BOUND_USABLE_TICK_MAGNITUDE } from "./twammPool";

function toSortedTicks(liquidity: bigint) {
  return [
    { tick: -MAX_BOUND_USABLE_TICK_MAGNITUDE, liquidityDelta: liquidity },
    { tick: MAX_BOUND_USABLE_TICK_MAGNITUDE, liquidityDelta: -liquidity },
  ];
}

describe("oraclePool", () => {
  describe("quote", () => {
    it("quote token0", () => {
      const pool = new OraclePool({
        token0: 0n,
        token1: 1n,
        sqrtRatio: toSqrtRatio(1),
        liquidity: 1000000000n,
        tick: 0,
        extension: 1n,
        sortedTicks: toSortedTicks(1000000000n),
        lastSnapshotTime: 0n,
      });

      const { executionResources, calculatedAmount } = pool.quote({
        tokenAmount: {
          amount: 1000n,
          token: 0n,
        },
        meta: { block: { number: 1, time: 32 } },
      });

      expect(calculatedAmount).toMatchSnapshot();
      expect(executionResources.initializedTicksCrossed).toEqual(0);
      expect(executionResources.snapshotUpdated).toEqual(true);
    });

    it("quote production example", () => {
      const pool = new OraclePool({
        token0: 0n,
        token1: 1n,
        sqrtRatio: 20685032069219899469794685553443565n,
        liquidity: 1825741529166n,
        tick: -19416252,
        extension: 1n,
        sortedTicks: toSortedTicks(1825741529166n),
        lastSnapshotTime: 0n,
      });

      const baseQuote = pool.basePool.quote({
        tokenAmount: {
          amount: 10n ** 6n,
          token: 1n,
        },
        meta: { block: { number: 1, time: 32 } },
      });

      expect(
        pool.quote({
          tokenAmount: {
            amount: 10n ** 6n,
            token: 1n,
          },
          meta: { block: { number: 1, time: 32 } },
        }),
      ).toEqual({
        ...baseQuote,
        executionResources: {
          ...baseQuote.executionResources,
          snapshotUpdated: true,
        },
        stateAfter: { ...baseQuote.stateAfter, lastSnapshotTime: 32n },
      });
    });
  });
});
