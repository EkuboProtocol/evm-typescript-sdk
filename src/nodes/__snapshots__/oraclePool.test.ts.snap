// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`oraclePool > quote > quote token0 1`] = `999n`;
